File: cpp/graph/shortest_path/SPFA.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

struct EdgeNode {
    int index;
    int weight;
    EdgeNode *next;

    EdgeNode(int index, int weight) : index(index), weight(weight), next(nullptr) {}
};

struct VertexNode {
    int value;
    EdgeNode *next;

    VertexNode(int value) : value(value), next(nullptr) {}
};

class SPFA {
public:
    vector<int> search(vector<VertexNode *> &graphs, int start) {
        vector<bool> visited(graphs.size(), false);
        queue<int> queue;
        queue.push(start);

        vector<int> result(graphs.size(), INT_MAX);
        result[start] = 0;

        int count = 0;
        while (!queue.empty()) {
            int index = queue.front();
            queue.pop();
            visited[index] = false;

            EdgeNode *edge = graphs[index]->next;
            while (edge) {
                int edgeIndex = edge->index;
                int edgeWeight = edge->weight;
                if (result[edgeIndex] > result[index] + edgeWeight) {
                    result[edgeIndex] = result[index] + edgeWeight;
                    if (!visited[edgeIndex]) {
                        queue.push(edgeIndex);
                        visited[edgeIndex] = true;
                        ++count;
                    }
                }
                edge = edge->next;
            }
            if (count >= (int) graphs.size()) {
                return vector<int>(0);
            }
        }
        return result;
    }
};

File: py/graph/shortest_path/SPFA.py
#!/usr/bin/env python
# -*- coding:gbk -*-

import sys
from typing import List

class EdgeNode:
    def __init__(self, index: int, weight: int):
        self.index, self.weight, self.next = index, weight, None

class VertexNode:
    def __init__(self, value: int):
        self.value, self.next = value, None

class SPFA:
    def search(self, graphs: List[VertexNode], start: int) -> List[int]:
        visited, queue = [False for _ in range(len(graphs))], [start]
        result = [sys.maxsize for _ in range(len(graphs))]
        result[start] = 0

        count = 0
        while queue:
            index = queue.pop(0)
            visited[index] = False

            edge = graphs[index].next
            while edge:
                edgeIndex, edgeWeight = edge.index, edge.weight
                if result[edgeIndex] > result[index] + edgeWeight:
                    result[edgeIndex] = result[index] + edgeWeight
                    if not visited[edgeIndex]:
                        queue.append(edgeIndex)
                        visited[edgeIndex] = True
                        count += 1
                edge = edge.next
            if count >= len(graphs):
                return []
        return result
